lib/TPP/GPU/LinalgToGpu.cpp

```diff
@@ -93,6 +94,200 @@ static bool supportsMMACompute(linalg::LinalgOp linalgOp) {
+                               linalg::LinalgOp consumer,
+                               memref::StoreOp rootStoreOp,
+                               ValueRange storeIndices,
+                               PatternRewriter &rewriter) {
+  Location loc = rootOp.getLoc();
+  auto rootOutput = rootOp.getDpsInits()[0];
+
+  // Insert fused eltwise ops before the store and later replace the store
+  // with a new result.
+  OpBuilder::InsertionGuard guard(rewriter);
+  rewriter.setInsertionPoint(rootStoreOp);
+
+  Operation *newStore = nullptr;
+  SmallVector<Value> operands;
+  if (structured_match::utils::isTwoDAddOp(consumer, &operands)) {
+    // Get the value to be added. Load the element first, if necessary.
+    auto addValue = operands[0] != rootOutput ? operands[0] : operands[1];
+    if (addValue.getType().isa<ShapedType>()) {
+      addValue = rewriter.create<memref::LoadOp>(loc, addValue, storeIndices)
+                     .getResult();
+    }
+    // Fuse the add into the matmul body.
+    auto addOp =
+        rewriter.create<arith::AddFOp>(loc, rootStoreOp.getValue(), addValue);
+    // Store the new result.
+    newStore = rewriter.replaceOpWithNewOp<memref::StoreOp>(
+        rootStoreOp, addOp.getResult(), rootOutput, storeIndices);
+  } else if (structured_match::utils::isTwoDReluOp(consumer, &operands)) {
+    // Must be a floating point type.
+    auto outputType = rootOutput.getType().cast<ShapedType>();
+    auto floatType = dyn_cast<FloatType>(outputType.getElementType());
+    if (!floatType)
+      return nullptr;
+    // Fuse the relu into the matmul body.
+    Value zeroFloat = rewriter.create<arith::ConstantFloatOp>(
+        loc, APFloat::getZero(floatType.getFloatSemantics()), floatType);
+    auto maxOp = rewriter.create<arith::MaximumFOp>(loc, rootStoreOp.getValue(),
+                                                    zeroFloat);
+    // Store the new result.
+    newStore = rewriter.replaceOpWithNewOp<memref::StoreOp>(
+        rootStoreOp, maxOp.getResult(), rootOutput, storeIndices);
+  } else {
+    // Not a fusable operation. Bail out.
+    return nullptr;
+  }
+
+  rewriter.eraseOp(consumer);
+
+  return newStore;
+}
+
+// Fuse elementwise consumers.
+// A naive fusion strategy that looks at the other operations after the root
+// linalg op and tries to fuse them.
+// Attemps bails on the first mismatch.
+// Returns updated store op or nullptr if the fusion fails.
+static Operation *fuseEltwiseConsumers(linalg::LinalgOp rootOp,
+                                       Operation *rootStoreOp,
+                                       ValueRange storeIndices,
+                                       PatternRewriter &rewriter) {
+  auto parentOp = rootOp->getParentOp();
+  auto rootOutput = rootOp.getDpsInits()[0];
+
+  // Traverse other ops within the same region and collect consumers.
+  SmallVector<linalg::LinalgOp> consumers;
+  Operation *nextOp = rootOp;
+  while ((nextOp = nextOp->getNextNode())) {
```

Comment:
Is this safe?, example:
```
matmul (%A, %B) (%C)
func (%A, %B, %C) // can write/read to all the operands
relu (%C)
```
If you fuse and the relu and the matml, you can get wrong results.

Reply:
There are various checks for that.
In general, I collect only linalgs and stop walking over ops any other operation.
Anyway, added a test case like your example.